File: solutions/Python/three_sum_hash_set.py
from typing import List


def three_sum_hash_set(nums: List[int]) -> List[List[int]]:
    nums.sort()
    result = []
    n = len(nums)

    for i in range(n - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue
        seen = set()
        j = i + 1
        while j < n:
            need = -(nums[i] + nums[j])
            if need in seen:
                triplet = [nums[i], need, nums[j]]
                if triplet not in result:
                    result.append(triplet)
            seen.add(nums[j])
            j += 1

    return result


print(three_sum_hash_set([-1, 0, 1, 2, -1, -4]))  # [[-1, -1, 2], [-1, 0, 1]]
print(three_sum_hash_set([0, 1, 1]))               # []
print(three_sum_hash_set([0, 0, 0]))               # [[0, 0, 0]]

File: solutions/Golang/three_sum_hash_set.go
package main

import (
	"fmt"
	"sort"
)

func threeSumHashSet(nums []int) [][]int {
	sort.Ints(nums)
	result := [][]int{}
	n := len(nums)

	for i := 0; i < n-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		seen := make(map[int]bool)
		for j := i + 1; j < n; j++ {
			need := -(nums[i] + nums[j])
			if seen[need] {
				triplet := []int{nums[i], need, nums[j]}
				duplicate := false
				for _, existing := range result {
					if existing[0] == triplet[0] && existing[1] == triplet[1] && existing[2] == triplet[2] {
						duplicate = true
						break
					}
				}
				if !duplicate {
					result = append(result, triplet)
				}
			}
			seen[nums[j]] = true
		}
	}

	return result
}

func main() {
	nums := []int{-1, 0, 1, 2, -1, -4}
	result := threeSumHashSet(nums)
	fmt.Println(result) // [[-1 -1 2] [-1 0 1]]
}
